tensorflow/lite/micro/examples/mnist_lstm/train.py

```diff
@@ -0,0 +1,166 @@
+        tf.keras.Model: A Keras LSTM model
+    """
+
+  model = tf.keras.models.Sequential([
+      tf.keras.layers.Input(shape=(28, 28), name='input'),
+      tf.keras.layers.LSTM(units, return_sequences=True),
+      tf.keras.layers.Flatten(),
+      tf.keras.layers.Dense(10, activation=tf.nn.softmax, name='output')
+  ])
+  model.compile(optimizer='adam',
+                loss='sparse_categorical_crossentropy',
+                metrics=['accuracy'])
+  model.summary()
+  return model
+
+
+def get_train_data():
+  """Get MNIST train and test data
+
+    Returns:
+        list of tuples: (data, label) pairs for train and test
+    """
+  (x_train, y_train), _ = tf.keras.datasets.mnist.load_data()
+  x_train = x_train / 255.  # normalize pixel values to 0-1
+  x_train = x_train.astype(np.float32)
+  return (x_train, y_train)
+
+
+def train_lstm_model(epochs):
+  """Train keras LSTM model on MNIST dataset
+
+    Args:
+        epochs (int) : number of epochs to train the model
+
+    Returns:
+        tf.keras.Model: A trained keras LSTM model
+    """
+  model = create_model()
+  x_train, y_train = get_train_data()
+  callback = tf.keras.callbacks.EarlyStopping(
+      monitor='val_loss',
+      patience=1)  #early stop if validation loss does not drop anymore
+  model.fit(x_train,
+            y_train,
+            epochs=epochs,
+            validation_split=0.2,
+            batch_size=32,
+            callbacks=[callback])
+  return model
+
+
+def save_tf_model(model, save_path):
+  """Save the trained LSTM model in tensorflow format
+
+    Args:
+        model (tf.keras.Model): the trained LSTM Model
+        save_path (str): path to save the model
+    """
+
+  batch_size, steps, input_size = 1, 28, 28
+  run_model = tf.function(lambda x: model(x))
+  concrete_func = run_model.get_concrete_function(
+      tf.TensorSpec([batch_size, steps, input_size], model.inputs[0].dtype))
+  model.save(save_path, save_format="tf", signatures=concrete_func)
+  print(f'TF model saved to {save_path}')
+
+
+def save_tflite_model(model, saved_path, optimize):
+  """Convert the saved TF model to tflite model, then save it as .tflite flatbuffer format
+
+    Args:
+        model (tf.keras.Model): the trained LSTM Model
+        save_path (str): path to save the model
+        optimize (bool): enable model conversion optimization (dynamic range quantization)
+    """
+  fixed_input = tf.keras.layers.Input(shape=[28, 28],
+                                      batch_size=1,
+                                      dtype=model.inputs[0].dtype,
+                                      name='fixed_input')
+  fixed_output = model(fixed_input)
+  run_model = tf.keras.models.Model(
+      fixed_input, fixed_output
+  )  #converter requires fixed shape input to work, alternative: b/225231544
+  converter = tf.lite.TFLiteConverter.from_keras_model(run_model)
+  save_name = 'lstm.tflite'
+  if optimize:
+    converter.optimizations = [tf.lite.Optimize.DEFAULT]
```

Comment:
TFLM doesn't support dynamic range quantization right now so I think LSTM is using another quantization scheme? Unfortunately I don't see it in the original colab so perhaps we want to take this out now and figure out the right thing to do when we add the interpreter script.

Reply:
Sorry I am new to quantization. I was following this: https://www.tensorflow.org/lite/performance/post_training_quantization#dynamic_range_quantization 

They listed it under dynamic range quantization but they stated 
"The simplest form of post-training quantization statically quantizes only the weights from floating point to integer, which has 8-bits of precision:" So I guess this is actually static weight int8 quantization?

Reply:
I think "statically quantizes" here probably just means PTQ and not during training. "At inference, weights are converted from 8-bits of precision to floating point and computed using floating-point kernels." Is that what we do for LSTM? Currently other ops typically do full integer quantization only.

Reply:
I think so. Should I call it full integer quantization then? How can I check it? Through the generated HTML visualization?

Reply:
No... full integer quantization requires a representative dataset, so what you're doing here is dynamic quantization. I just don't know which is supported by LSTM. I guess the easiest way to find out is just by running it :grimacing:

Reply:
I will keep the option of generating the quantized model. Instead of saying it is doing dynamic quantization, I am going to print `Model conversion using tf.lite.Optimize.DEFAULT` instead

Reply:
let's remove the quantization for now. It is indeed going to produce a model that is dynamic range quantized and we won't be able to run inference with the TFLM interpreter so that would be confusing.

Let's start with floating point inference and add a fully quantized model with a representative dataset in a follow up PR. We'll want to start looking into QAT etc. as well.